packages/flutter_tools/lib/src/run.dart

```diff
@@ -311,78 +323,82 @@ class RunAndStayResident {
-        return null;
+        _devFS = null;
+        printError('Error initializing DevFS: $error');
+        return false;
       }
-    }
 
-    // Send the root and lib directories.
-    Directory directory = Directory.current;
-    _sendFiles(directory, '', _dartFiles(directory.listSync()));
-
-    directory = new Directory('lib');
-    _sendFiles(directory, 'lib', _dartFiles(directory.listSync(recursive: true)));
-
-    // Send the packages.
-    if (FileSystemEntity.isFileSync(kPackagesFileName)) {
-      PackageMap packageMap = new PackageMap(kPackagesFileName);
-
-      for (String packageName in packageMap.map.keys) {
-        Uri uri = packageMap.map[packageName];
-        // Ignore self-references.
-        if (uri.toString() == 'lib/')
-          continue;
-        Directory directory = new Directory.fromUri(uri);
-        if (directory.existsSync()) {
-          _sendFiles(
-            directory,
-            'packages/$packageName',
-            _dartFiles(directory.listSync(recursive: true))
-          );
-        }
-      }
+      _exitCompleter.future.then((_) async {
+        await _cleanupDevFS();
+      });
     }
 
-    try {
-      await devFS.flush();
-    } catch (error) {
-      printError('Error sending sources to the client device: $error');
-    }
+    printStatus('DevFS: Updating files on device...');
+    await _devFS.update();
+    printStatus('DevFS: Finished updating files on device...');
+    return true;
   }
 
-  void _sendFiles(Directory base, String prefix, List<File> files) {
-    String basePath = base.path;
-
-    for (File file in files) {
-      String devPath = file.path.substring(basePath.length);
-      if (devPath.startsWith('/'))
-        devPath = devPath.substring(1);
-      devFS.stageFile(prefix.isEmpty ? devPath : '$prefix/$devPath', file);
+  Future<Null> _cleanupDevFS() async {
+    if (_devFS != null) {
+      // Cleanup the devFS.
+      await _devFS.destroy();
     }
+    _devFS = null;
   }
 
-  List<File> _dartFiles(List<FileSystemEntity> entities) {
-    return new List<File>.from(entities
-      .where((FileSystemEntity entity) => entity is File && entity.path.endsWith('.dart')));
+  Future<Null> _launchFromDevFS(ApplicationPackage package,
+                                String mainScript) async {
+    String entryPath = path.relative(mainScript, from: _devFSProjectRootPath);
+    String deviceEntryPath =
+        _devFS.baseUri.resolve(entryPath).toFilePath();
+    String devicePackagesPath =
+        _devFS.baseUri.resolve('.packages').toFilePath();
+    await device.runFromFile(package,
+                             deviceEntryPath,
+                             devicePackagesPath);
+  }
+
+  Future<bool> _reloadSources() async {
+    if (observatory.firstIsolateId == null)
+      throw 'Application isolate not found';
+    if (_devFS != null) {
+      await _updateDevFS();
+    }
+    printStatus('HOT: Performing hot reload.');
```

Comment:
Let's change this to:

```
Status status = logger.startProgress('Performing hot reload');
```

and end it with

```
status.stop(showElapsedTime: true);
```

That will show a spinner on the command line.